sdk/python/src/dagger/client/_guards.py

```diff
@@ -58,6 +59,20 @@ def typecheck(func: Callable[_P, _T]) -> Callable[_P, _T]:
     ...
 
 
+@overload
+def typecheck(
+    func: Callable[Concatenate[_S, _P], Coroutine[Any, Any, _T]]
+) -> Callable[Concatenate[_S, _P], Coroutine[Any, Any, _T]]:
+    ...
```

Comment:
This is still not quite (Py)right 😄:

<img width="591" alt="Screenshot 2024-02-18 at 22 52 26" src="https://github.com/dagger/dagger/assets/174525/0d425973-5af1-4686-bfe3-ba46a6161d1a">


The `typecheck` decorator only exists to customize the error message from [beartype](https://pypi.org/project/beartype/), however, [v0.17.0](https://github.com/beartype/beartype/releases/tag/v0.17.0) seems to “massively increase the configurability of @beartype”, so I have it on my backlog to see if we can remove the wrapped decorator now.

You can see that it solves your issue if you replace all lines since `_T` with just:
```python
typecheck = beartype
```

So, to approve this PR, I either need to see the overloads pass on Pyright, or better, configure beartype so that we don’t have to wrap it.